src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -173,25 +157,37 @@ void ExpressionAnalyzer::analyzeAggregation()
 
     if (select_query)
     {
+        NamesAndTypesList array_join_columns;
+
         bool is_array_join_left;
-        ASTPtr array_join_expression_list = select_query->arrayJoinExpressionList(is_array_join_left);
-        if (array_join_expression_list)
+        if (ASTPtr array_join_expression_list = select_query->arrayJoinExpressionList(is_array_join_left))
         {
             getRootActionsNoMakeSet(array_join_expression_list, true, temp_actions, false);
-            addMultipleArrayJoinAction(temp_actions, is_array_join_left);
+            if (auto array_join = addMultipleArrayJoinAction(temp_actions, is_array_join_left))
+            {
+                auto sample_block = temp_actions->getSampleBlock();
+                array_join->prepare(sample_block);
+                temp_actions = std::make_shared<ExpressionActions>(sample_block.getColumnsWithTypeAndName(), context);
+            }
 
-            array_join_columns.clear();
             for (auto & column : temp_actions->getSampleBlock().getNamesAndTypesList())
                 if (syntax->array_join_result_to_source.count(column.name))
                     array_join_columns.emplace_back(column);
         }
 
+        columns_after_array_join = sourceColumns();
+        columns_after_array_join.insert(columns_after_array_join.end(), array_join_columns.begin(), array_join_columns.end());
+
         const ASTTablesInSelectQueryElement * join = select_query->join();
         if (join)
         {
             getRootActionsNoMakeSet(analyzedJoin().leftKeysList(), true, temp_actions, false);
             addJoinAction(temp_actions);
         }
+
+        columns_after_join = columns_after_array_join;
+        const auto & added_by_join = analyzedJoin().columnsAddedByJoin();
+        columns_after_join.insert(columns_after_join.end(), added_by_join.begin(), added_by_join.end());
```

Comment:
Big `ARRAY JOIN` was always added before join (look at deleted `addMultipleArrayJoinAction` few lines before).
It used only columns from left table and had no information about columns from right. So, it could not have worked that way.
Function `arrayJoin` is still into actions, it shouldn't be broken.